### DISCUSSION TITLE
Script Web section: Use modulo % within %if ... else Conditional Statements

### DISCUSSION THREAD

**Discussion by ottelo9** [arendst/Tasmota#15241] [Q&A] [answered]
I try to use modulo % inside a conditional statement inside my html code:

```
%for tmp 1 s4h[-1] 1
%if (tmp-1)%4==0
%then
$['%(s(1.0((tmp-1)/4))+" min")%',%s4h[tmp]%],
%else
$['',%s4h[tmp]%],
%endif
%next
```

It seems to work but I get lot of outputs on my html page:

```
(tmp-1)???
???
???
???
(tmp-1)???
???
???
???
...
```

What can I do? To add another % to % (%%) doesnt help!

**Comment by gemu2015** [comment, marked as answer]
inside %% only a single string variable is supported
inside %()% only numeric formulas are supported.
in general i would not recommend to solve dynamic html that way.

try this method
wcs sends the html code  (wcs = web content send)
#tsub
if wm==0 {
for tmp 1 s4h[-1] 1
wcs html code
next
}

\>W
%=#tsub

here is an example of a complex dynamic html script

[web_gui_setup.txt.zip](https://github.com/arendst/Tasmota/files/8364861/web_gui_setup.txt.zip)

<img width="715" alt="Bildschirmfoto 2022-03-28 um 19 11 32" src="https://user-images.githubusercontent.com/11647075/160451520-5d62d724-fbca-4012-a9c5-d5541b11bf96.png">

 
<img width="381" alt="Bildschirmfoto 2022-03-28 um 19 12 17" src="https://user-images.githubusercontent.com/11647075/160451530-c72fff2d-0e94-4f1d-89dd-9fef83fbcb89.png">

**Reply by ottelo9** [reply]
THX :)

**Reply by ottelo9** [reply]
@gemu2015 
Mhh this doesnt work! The array for google chart consists only of the header ['t', 'Leistung [W]']. The data from the sub is missing.

**My web section:**
```
$<br><div id="customchart1"style="width:640px;height:480px;margin:0 auto"></div><br>
$<script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
$<script type="text/javascript">google.charts.load('current',{packages:['corechart']});</script>
$<script language="JavaScript">function drawChart(){
$var data=google.visualization.arrayToDataTable([
$['t', 'Leistung [W]'],
%=#wsub
$]);
$var options = {
$chartArea:{left:60,width:'83%%'},
$width:'100%%',
$legend: 'none',
$title:'Leistungsverlauf 4h [Watt]',
$hAxis:{direction:-1}
$};
$var chart=new google.visualization.LineChart(document.getElementById('customchart1'));
$chart.draw(data,options);}
$google.charts.setOnLoadCallback(drawChart);</script>
```

**my sub:**
```
#wsub
if (wm==0) {
	for tmp 1 s4h[-1] 1
		if ((tmp-1)%4==0) {
			wcs $['%(s(1.0((tmp-1)/4))+" min")%',%s4h[tmp]%]
		} else {
			wcs $['',%s4h[tmp]%],
		}
	next
}
```

**Reply by gemu2015** [reply]
webmode is set in >W section e.g. by the $ sign.
in wsub you have to modify
if (wm>0) { ...
and wcs without the preceding $ char

btw why aren't you using the build in google charts ?

**Reply by ottelo9** [reply]
What do you mean „webmode set“? I doesn’t found any info about that. 

I using my own google chart code because I want a custom time x-axis (I wrote you a PM in the creatix Forum). From now to the past (-4hours). One value every 15s. Do you know how to calc a correct timestamp for the x axis? From now to past (-4h) in 15s steps. In C I would use utc time in seconds and substract 15s.

**Reply by gemu2015** [reply]
from the docs:
wm = contains source of web request code e.g. 0 = Sensor display (FUNC_WEB_SENSOR)

webmode is set by certain preceding characters in the >W section. $ sets webmode to 36

you can access the standard time variables:
hours = hours
mins = mins
secs = seconds

(also check the creatix forum)

**Reply by ottelo9** [reply]
lol simply remove the $. Thank you. I've been trying for hours.

**Reply by ottelo9** [reply]
I know how to get the actual hours, mins and secs. But its very laborious to calculate the timestring for every entry instead of simply put a UTC time to the google chart table. This is my sub to generate the time string. Takes some time for 960 values...:

```
#calct
scs-=15
if (scs<0)
then
  scs=45
  mns-=1
  if (mns<0)
  then
    mns=59
    hrs-=1
    if (hrs<0)
	then
      hrs=23
    endif
  endif
endif
cstrr=s(2.0hrs)+":"+s(2.0mns)
```

**Reply by gemu2015** [reply]
for such a large array i doubt to directly use a script.
anyhow i added a cmd s2hms to convert seconds to HH:MM:SS
this should speed up a bit.
this also avoids a subroutine which adds some time too.

[xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/8390493/xdrv_10_scripter.ino.zip)

**Reply by ottelo9** [reply]
Thanks! Now I can put the seconds (get minutes with "time" * 60 = seconds) in your new cmd. THX!!

**Reply by ottelo9** [reply]
@gemu2015 
Do you want to commit it? I try to test it now.

**Reply by ottelo9** [reply]
Runs very fast!!

**Reply by ottelo9** [reply]
@gemu2015 
How can I refresh the chart? I tried to put the sub into the >S section (every tele period), but this dont work. I dont understand the wm thing.

**Reply by gemu2015** [reply]
no auto update to google charts currently implemented. you may refresh the page with browser or on regular intervals with
e.g every 5 seconds:

$<script> setTimeout("location.reload(true);",5000); </script>

**Reply by ottelo9** [reply]
@gemu2015 Do you want to commit it? s2hms()

**Reply by gemu2015** [reply]
Already merged
